llvm/lib/Transforms/Utils/InlineFunction.cpp

```diff
@@ -1477,6 +1483,12 @@ static void AddReturnAttributes(CallBase &CB, ValueToValueMapTy &VMap) {
     auto *NewRetVal = dyn_cast_or_null<CallBase>(VMap.lookup(RetVal));
     if (!NewRetVal)
       continue;
+
+    // The RetVal might have be simplified during the inlining
+    // process. Only propagate return attributes if we are in fact calling the
+    // same function.
+    if (RetVal->getCalledFunction() != NewRetVal->getCalledFunction())
```

Comment:
This doesn't seem like it will reliably detect the issue.

First off, if both calls are indirect, both calls to getCalledFunction() would return nullptr, in which case you aren't getting any useful information at all.

Second, even if both calls are calling the "same" function, it might not have the same behavior if the arguments are different.  Or global state is different.  Not sure how likely this is given the limited simplifications the inline performs, but it's hard to rule out.

So I think we need to avoid using VMap, and pass down the equivalence information some other way.

Reply:
> This doesn't seem like it will reliably detect the issue.
> 
> First off, if both calls are indirect, both calls to getCalledFunction() would return nullptr, in which case you aren't getting any useful information at all.
> 

Fair, I think its fine to just disable all prop for indirect functions.

> Second, even if both calls are calling the "same" function, it might not have the same behavior if the arguments are different. Or global state is different. Not sure how likely this is given the limited simplifications the inline performs, but it's hard to rule out.
> 
> So I think we need to avoid using VMap, and pass down the equivalence information some other way.

Is that really a valid concern? We don't allow arbitrary transforms on the new instructions. Within the confines of what we do do, I don't see how we could change global state or produce and argument that is not semantically equiv to what was in the original.

That being said, we could also move the propagation code to the function cloning logic before simplifications/etc.

Reply:
Suppose you inline a function that contains two calls to foo().  The second call gets simplified away, and the simplified value is the first call to foo().  So the vmap lookup finds the first call to foo(), which has completely different arguments/state.

Reply:
You are worried about something like:
```
declare ptr @foo(ptr) memory(none) willreturn nounwind

define ptr @caller(ptr %p, i1 %c) {
  %r = call ptr @callee(ptr %p, i1 %c)
  ret ptr %r
}

define ptr @callee(ptr %p, i1 %c) {
  %x = call ptr @foo(ptr %p)
  br i1 %c, label %T, label %F
T:
  ret ptr %x
F:
  %r2 = call nonnull ptr @foo(ptr %p)
  ret ptr %r2
}
```
?

Where we might prop the `nonnull` from the second call to the first call if it got simplified?

Reply:
I guess the other options are either move this to before simplifications in the `cloneFunction` call or track which callbases get simplified and skip them.

Reply:
Or we could just disable the simplifyInstruction on callbases

Reply:
Personally I'm for just not using simplifyInstruction on callbases in CloneFunction in inliner. Don't see much value in it in a realistic pipeline and unless I'm missing something, I think it solves the bug.
But would like to hear if anyone else has an opinion on how this should be solved.